### ISSUE TITLE
Virtual gamepads disappeared in 1.4.0

### ISSUE THREAD

**Issue by MarioFlying** [Swordfish90/Lemuroid#62]
On-screen controls aren't shown in any game or system since I updated to the latest version. Before 1.4.0, on-screen controls were correctly displayed.
![Screenshot_2020-05-12-12-59-49-967_com swordfish lemuroid](https://user-images.githubusercontent.com/2698205/81677549-55256f80-9451-11ea-95bd-ec57e53c86c6.jpg)

**Comment by Swordfish90** [comment]
Hi, I'm sorry you are experiencing this. In version 1.4.0 there were very limited changes to layouts, but on version 1.3.0 there were basically redone. Is it possible that you skipped an update?

A few questions to better understand the problem: which device are you using? Are the layouts missing in any console? Do you have a gamepad connected?

**Comment by MarioFlying** [comment]
I'm on a Xiaomi Redmi Note 8 Pro, Android 9, MIUI 11.3.

Yes, this happens with every console.

I seem to have a similar problem in a racing game too (Horizon Chase) where on every race the external control scheme is selected and I have to switch back, even if Bluetooth is turned off from the start.  Seems like there's some ghost gamepad taking control, must investigate further. Unfortunately, Lemuroid is now broken for me unless I play with a physical gamepad.

**Comment by MarioFlying** [comment]
And yes, I did skip version 1.3.0, so it likely happened on that version. I'll try it out and confirm.

I did a search and found out a 2015 report from a guy using an integrated controller having the same issue with Horizon Chase; turns out the game uses Unity's controller routines, which hide on-screen controls automatically as soon as a physical controller is detected, instead of relying on the user's manual selection. Maybe it's a similar issue?

**Comment by MarioFlying** [comment]
Nope. I confirm that 1.3.0 works perfectly.

The problem is with 1.4.0, confirmed.

![Screenshot_2020-05-24-12-37-49-702_com swordfish lemuroid](https://user-images.githubusercontent.com/2698205/82752078-ef73a480-9dbb-11ea-9432-78d97acdb8a2.jpg)
![Screenshot_2020-05-24-12-39-26-646_com swordfish lemuroid](https://user-images.githubusercontent.com/2698205/82752105-16ca7180-9dbc-11ea-8bb7-e9e8ec235177.jpg)
![Screenshot_2020-05-24-12-39-15-777_com miui securitycenter~01](https://user-images.githubusercontent.com/2698205/82752112-1b8f2580-9dbc-11ea-85e7-ba7903ce5117.jpg)

**Comment by RandomHam** [comment]
Hey, I got the same Issue when installing 1.4 clean or after upgrading to the current beta via play store. My device is a reminder note 9s

**Comment by Swordfish90** [comment]
Could you please check the 1.6.0-beta1 which I just published on the beta channel? It contains a tentative fix for the issue.

**Comment by RandomHam** [comment]
Hey, sorry for the late response. I'd love to test it but I'm not quite sure what you mean, I'm sorry. Is this version supposed to be in separate GitHub repo or branch? If its supposed to be on the play store beta version, I haven't received it yet.

**Comment by Swordfish90** [comment]
@RandomHam sorry, I wasn't really clear. I meant the version published as beta on Google Play (I still haven't tagged the release here on github). Sometimes it takes a few hours before it reaches every user, so that's expected.

**Comment by RandomHam** [comment]
no worries and thanks for the clear up. I have gotten the update and it definetly works now and is useable quite well.

There is only a minor issue of which I don't know whether it is related to the the changes you made or has been there before: the screen orientation of the app while playing only changes after visiting the "edit controls" menu

Steps to reproduce:

1. start lemuroid
2. start game/rom
3. press cog icon on virtual gamepad
4. press on "Edit controls"
5. close the popup

only after those steps the  game constantly adjust its orientation with me roatating the phone not before and not after reopening the game from home (just using the back button). The gamepad keeps canging as well as intended btw, let me know if I should post that as a separate issue

**Comment by Swordfish90** [comment]
@RandomHam ... Thank you for the very quick validation. I guess we can close the issue.

Regarding the other problem you spotted I've been able to reproduce it, but only when the device orientation is locked. Can you check if you have forced the device to be portrait only? (In that case the bug becomes, "the screen rotates even if I force portrait orientation").

**Comment by RandomHam** [comment]
Yes, I have auto-rotate disabled. I did not even think of that being the cause since never use auto-rotate. 

Having the option to manually set the orientation disregarding the phones' rotation would be a nice additional feature imo :P

**Comment by Swordfish90** [comment]
Apparently this issue is still reproducible on some device which report the fingerprint scanner as a gamepad. It's clearly a manufacturer issue, but we should probably add a way to disable the default behaviour.

https://github.com/libgdx/libgdx/issues/5596

**Comment by MarioFlying** [comment]
The issue was fixed for me after upgrading to MIUI 12 with Android 10. Sorry for the late response.

**Comment by RandomHam** [comment]
The issue re-emerged for me after the last update (2nd Octobre) via the play store beta program, Controls are missing again and the game screen is automatically centered (as opposed to sticking to the top)

@Swordfish90 You might be right that the fingerprint sensor ist causing the issue. I have a bug where touching it results "screen-press aimed at the top-most visible button on the screen" (hope that's understandable ...). 
I'm using a custom rom which I'm planning to update (including the firmware) next week when find time for that, maybe that fixes it for me.

**Comment by Swordfish90** [comment]
Summaring the issue, it seem we cannot trust Android correctly assigning ports (as we recently saw with the Xiaomi MiBox), nor the gamepads correctly reporting the right features. I just merged #111 , so we went back to manually manage gamepads, but I also added a switch on mobile to never hide touch controls (this should fix the fingerprint scanner that behaves like a gamepad).

I'll publish everything in beta in a day or two, so please let me know if you're still having issues with gamepads not recognized, things recognized as gamepads and missing touch controls.

**Comment by RandomClaim** [comment]
Got the update, touch controls are working for me again. At least in gba and nds, haven't tester others

**Comment by Swordfish90** [comment]
Thank you very much @RandomClaim. The code is exactly the same, so if it's working for one, it should be working for every console.

All the changes will be on beta in a few hours, so I'm going to close the issue.

Feel free to reopen it if you're still able to reproduce it and we'll look into that specifically.
